src/services/suggestionDiagnostics.ts

```diff
@@ -156,7 +156,29 @@ namespace ts {
     }
 
     function isPromiseHandler(node: Node): node is CallExpression {
-        return isCallExpression(node) && (hasPropertyAccessExpressionWithName(node, "then") || hasPropertyAccessExpressionWithName(node, "catch"));
+        return isCallExpression(node) && (
+            hasPropertyAccessExpressionWithName(node, "then") && hasZeroOrOneNonNullArguments(node) ||
+            hasPropertyAccessExpressionWithName(node, "catch"));
+    }
+
+    function hasZeroOrOneNonNullArguments(node: CallExpression) {
+        if (node.arguments.length < 2) return true;
+        let seenArg = false;
+        for (const arg of node.arguments) {
+            switch (arg.kind) {
+                case SyntaxKind.NullKeyword:
+                    continue;

```

Comment:
Why can't you just break the loop at this point? Is there a three argument version?

Reply:
I had the same question and my guess was `then(undefined, handler)`.

Reply:
That was in the back of my mind, but it makes more sense to disable the refactor if there are more than two arguments since we can’t understand what that’s supposed to mean.